qiskit_experiments/calibration/calibrations.py

```diff
@@ -0,0 +1,419 @@
+# that they have been altered from the originals.
+
+"""Class to store the results of a calibration experiments."""
+
+import copy
+import dataclasses
+from collections import namedtuple
+from datetime import datetime
+from typing import Any, Dict, Set, Tuple, Union, List, Optional
+
+from qiskit.circuit import Gate
+from qiskit import QuantumCircuit
+from qiskit.pulse import Schedule, DriveChannel, ControlChannel, MeasureChannel
+from qiskit.pulse.channels import PulseChannel
+from qiskit.circuit import Parameter
+from qiskit_experiments.calibration.exceptions import CalibrationError
+from qiskit_experiments.calibration.parameter_value import ParameterValue
+
+ParameterKey = namedtuple("ParameterKey", ["schedule", "parameter"])
+
+
+class Calibrations:
+    """
+    A class to manage schedules with calibrated parameter values.
+    Schedules are stored in a dict and are intended to be fully parameterized,
+    including the index of the channels. The parameter values are stored in a
+    dict where parameters are keys.
+    """
+
+    def __init__(self, backend):
+        """
+        Initialize the instructions from a given backend.
+
+        Args:
+            backend: The backend from which to get the configuration.
+        """
+
+        self._n_qubits = backend.configuration().num_qubits
+        self._n_uchannels = backend.configuration().n_uchannels
+        self._properties = backend.properties()
+        self._config = backend.configuration()
+        self._params = {}
+
+        # Required because copying a template schedule creates new paramters with new IDs.
+        self._parameter_map = {}
+        self._schedules = {}
+
+    def add_schedules(self, schedules: Union[Schedule, List[Schedule]]):
+        """
+        Add a schedule and register the parameters.
+
+        Args:
+            schedules: The schedule to add.
+
+        Raises:
+            CalibrationError: If the parameterized channel index is not formatted
+                following index1.index2...
+        """
+        if isinstance(schedules, Schedule):
+            schedules = [schedules]
+
+        for schedule in schedules:
+
+            # check that channels, if parameterized, have the proper name format.
+            # pylint: disable = raise-missing-from
+            for ch in schedule.channels:
+                if isinstance(ch.index, Parameter):
+                    try:
+                        [int(index) for index in ch.index.name.split(".")]
+                    except ValueError:
+                        raise CalibrationError(
+                            "Parameterized channel must have a name "
+                            "formatted following index1.index2..."
+                        )
+
+            self._schedules[schedule.name] = schedule
+
+            for param in schedule.parameters:
+                self._parameter_map[ParameterKey(schedule.name, param.name)] = param
```

Comment:
Perhaps we can make this easier by adding a flag which makes the decision a conscious one? E.g. change the signature of the method to:
```python
def add_schedules(self, schedules: Union[Schedule, List[Schedule]], override_existing: bool = False):
```
Then if the schedule name is already present and `override_existing` is `False` we raise an error. We could also check the hash of the parameters for existing schedule to avoid history loss. How does that sound?

Reply:
Depending on the work flow, it might be sufficient to forbid a schedule to be added twice. If it is easy to save to and reload from the database, the answer might be that one should create a new `CalibrationsDefinition` instance and reload if one wants to redefine schedules. On the other hand, maybe it is something that a user would expect to be able to do, so it should be allowed.

I don't think it needs by default to be an error for a user to redefine a schedule. It is when a user replaces a `Parameter` that I was concerned about. The storage of parameters is `(schedule.name, parameter.name): Parameter` in `_parameter_map` and `Parameter: {qubits: [ParameterValue, ...]}` in `_params`. So if `(schedule.name, parameter.name)` is updated to point to a new `Parameter` in `_parameter_map`, that `(schedule.name, parameter.name)` loses the history stored in the old `Parameter` entry in `_params`. I think this is how it worked when I made the comment above. Now with the way you wrote `register_parameter`, a `CalibrationError` will be raised for the second schedule if it reuses any of the parameter names of the first. `register_parameter(p, s)` probably should not raise an exception if `p == _parameter_map[(s.name, p.name)]` (re-adding a schedule with the same parameter instances).

What did you mean by "check the hash?" Do you mean that if `(schedule.name, parameter.name)` is already in `_parameter_map` that `_params[_parameter_map[(schedule.name, parameter.name)]]` would be copied to `_params[parameter]` if `parameter != _parameter_map[(schedule.name, parameter.name)]`? That seems like a natural thing to do. It allows for messy situations though where `(xp, amp)` and `(yp, amp)` point to the same `Parameter` in `_params` but then calling `add_schedules(xp)` a second time with a new `amp` Parameter results in `xp` and `yp` pointing to separate entries in `_params`.

I think this is somewhat of an edge case concern, but it is a question of how much flexibility to give the user vs keeping the implementation simple and giving the user less chance to get confused.

Reply:
Part of this is addressed in 72ce1c0. We still have the issue of tracking the history of the links between parameters and schedules when schedule are re-added.